pkg/doc/verifiable/credential.go

```diff
@@ -316,6 +378,128 @@ func NewCredential(dataJSON []byte, opts ...CredentialOpt) (*Credential, error)
 	return cred, nil
 }
 
+// credentialClaims is JWT Claims extension by Credential.
+type credentialClaims struct {
+	*jwt.Claims
+
+	Credential *rawCredential `json:"vc,omitempty"`
+}
+
+// rawCredentialClaims is used to get raw content of "vc" claim of JWT.
+type rawCredentialClaims struct {
+	*jwt.Claims
+
+	Raw map[string]interface{} `json:"vc,omitempty"`
+}
+
+func decodeJWT(rawJwt []byte, fetcher PublicKeyFetcher) ([]byte, *rawCredential, error) {
+	parsedJwt, err := jwt.ParseSigned(string(rawJwt))
+	if err != nil {
+		return nil, nil, fmt.Errorf("VC is not JWS: %w", err)
+	}
+
+	credClaims := new(credentialClaims)
+	err = parsedJwt.UnsafeClaimsWithoutVerification(credClaims)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse JWT claims: %w", err)
+	}
+
+	if verifyErr := verifyJWTSignature(parsedJwt, fetcher, credClaims); verifyErr != nil {
+		return nil, nil, verifyErr
+	}
+
+	// Apply VC-related claims from JWT.
+	credClaims.refineVCFromJWTClaims()
+
+	// Decode again to get raw content of "vc" claim.
+	rawClaims := new(rawCredentialClaims)
+	err = parsedJwt.UnsafeClaimsWithoutVerification(rawClaims)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse JWT claims: %w", err)
+	}
+
+	// Complement original "vc" JSON claim with data refined from JWT claims.
+	rawClaims.mergeRefinedVC(credClaims.Credential)
+
+	var vcData []byte
+	if vcData, err = json.Marshal(rawClaims.Raw); err != nil {
+		return nil, nil, errors.New("failed to marshal 'vc' claim of JWT")
+	}
+
+	return vcData, credClaims.Credential, nil
+}
+
+func (credClaims *credentialClaims) refineVCFromJWTClaims() {
+	raw := credClaims.Credential
+
+	if iss := credClaims.Issuer; iss != "" {
+		refineVCIssuerFromJWTClaims(raw, iss)
+	}
+
+	if nbf := credClaims.NotBefore; nbf != nil {
+		nbfTime := nbf.Time().UTC()
+		raw.Issued = &nbfTime
+	}
+
+	if jti := credClaims.ID; jti != "" {
+		raw.ID = credClaims.ID
+	}
+
+	if iat := credClaims.IssuedAt; iat != nil {
+		iatTime := iat.Time().UTC()
+		raw.Issued = &iatTime
+	}
+
+	if exp := credClaims.Expiry; exp != nil {
+		expTime := exp.Time().UTC()
+		raw.Expired = &expTime
+	}
+}
+
+func refineVCIssuerFromJWTClaims(raw *rawCredential, iss string) {
+	// Issuer of Verifiable Credential could be either string (id) or struct (with "id" field).
+	switch issuer := raw.Issuer.(type) {
+	case string:
+		raw.Issuer = iss
+	case map[string]interface{}:
+		issuer["id"] = iss
+	}
+}
+
+func (rawClaims *rawCredentialClaims) mergeRefinedVC(raw *rawCredential) {
+	rawVCClaims := rawClaims.Raw
+
+	// these operations are safe
+	rawData, _ := json.Marshal(raw) // nolint:errcheck
```

Comment:
@troyronda @llorllale @fqutishat are we going to ignore marshalling errors ? We had this discussion during daily Aries call today.

Reply:
I'm not understanding why marshal errors are being ignored.

Reply:
agree

Reply:
The takeaway was to forgive the drop to test coverage rather than ignore the errors altogether (or panic), and we aim for at least 85% to give ourselves a margin for cases like this. If some error should never happen, you don't want that error to be silent if the impossible ever happens because then you might be processing spurious data.

Reply:
I fixed that in #382